Make regression tests work when cvc5 is not available

tautschnig opened the pull request:
We should not have cvc5 as firm build (and test) dependency as CBMC works just fine when cvc5 is not available.

Please review commit-by-commit.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Make regression tests work when cvc5 or z3 is not available

We should not have cvc5 or z3 as firm build (and test) dependencies as
CBMC works just fine when these solvers are not available.

For cbmc-incr-smt2, both the z3 and cvc5 test runs are now conditional
on the respective solver being available on the PATH.

For cbmc-output-file, test descriptors are tagged with 'cvc5' or 'z3'
to indicate which solver they require. The Makefile dynamically excludes
tests for unavailable solvers using test.pl's -X flag, rather than
skipping the entire test suite.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

thomasspriggs commented:
A few thought/questions -

1. Could this give the false impression that the `cbmc` + `cvc5` tests have passed when they haven't been run? It would kind of make sense to me if this issue was handled using an option to exclude tests supplied by the build / test environment rather than something which is autodetected by the test suite.
2. Presumably this isn't needed for any `ctest` based test running because `ctest` can already be instructed to exclude a set of tests?
3. Why does this apply to `cvc5` but not `z3`? I can see how if the same logic is applied to both then there could be issues where the tests "pass" but `cbmc` hasn't been tested with any SMT solver. But I don't really see that the two solvers should be treated differently for testing when they aren't treated specially by the implementation.

When I was was working on the implementation I wanted to ensure that we tested with at least 2 solvers as insurance against accidentally producing an implementation which only works with one specific solver.

tautschnig commented:
> A few thought/questions -
> 
Thank you for your comments! A couple of notes:

>     1. Could this give the false impression that the `cbmc` + `cvc5` tests have passed when they haven't been run? It would kind of make sense to me if this issue was handled using an option to exclude tests supplied by the build / test environment rather than something which is autodetected by the test suite.
> 
Yes, at bare minimum I should add some output that tests have been skipped.

>     2. Presumably this isn't needed for any `ctest` based test running because `ctest` can already be instructed to exclude a set of tests?
> 
Good idea that we should perhaps not bake this into the CMake/ctest set-up and instead document how to skip those. (But we also need to make sure they are skippable at that level.)

>     3. Why does this apply to `cvc5` but not `z3`? I can see how if the same logic is applied to both then there could be issues where the tests "pass" but `cbmc` hasn't been tested with any SMT solver. But I don't really see that the two solvers should be treated differently for testing when they aren't treated specially by the implementation.
> 
See my PR notes: this should also apply to z3, I just haven't yet set aside the time to do it.

thomasspriggs commented:
> Good idea that we should perhaps not bake this into the CMake/ctest set-up and instead document how to skip those. (But we also need to make sure they are skippable at that level.)

The existing `ctest` argument you are looking for is `-E cvc5`.

tautschnig commented:
> > Good idea that we should perhaps not bake this into the CMake/ctest set-up and instead document how to skip those. (But we also need to make sure they are skippable at that level.)
> 
> The existing `ctest` argument you are looking for is `-E cvc5`.

Thanks, this will work for `cvc5`! But I don't think the same would be true of `z3`, which is used by tests spread out across several suites.

thomasspriggs commented:
> > > Good idea that we should perhaps not bake this into the CMake/ctest set-up and instead document how to skip those. (But we also need to make sure they are skippable at that level.)
> > 
> > 
> > The existing `ctest` argument you are looking for is `-E cvc5`.
> 
> Thanks, this will work for `cvc5`! But I don't think the same would be true of `z3`, which is used by tests spread out across several suites.

The `-E` argument does regex match against the test names provided to `add_test_pl_profile`. It could be made to work against the `z3` tests by adding "z3" to the names. For example by renaming `cbmc-new-smt-backend` to `cbmc-new-smt-backend-z3`. You can check the match results using the `-R` positive match argument with a dry run. For example `ctest -L CORE -V -N -R smt` to get all the "smt" related tests.

tautschnig committed:
Make CMake test suites conditional on solver availability

Use find_program() to detect z3 and cvc5 at CMake configure time and
only add test profiles that require these solvers when they are actually
available. This affects:

- cbmc-incr-smt2: z3 and cvc5 test profiles
- cbmc-output-file: tag-based exclusion of z3/cvc5 tests
- cbmc: incremental SMT2 backend test profile
- book-examples: incremental SMT2 backend test profile
- cbmc-primitives: fix bare message() call that would print
  Z3_EXISTS-NOTFOUND when z3 is not available

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
CI: run check-ubuntu-22_04-cmake-gcc without z3 and cvc5

Remove z3 and cvc5 from the check-ubuntu-22_04-cmake-gcc CI job to
verify that the test suite works correctly when these solvers are not
available. Other CI jobs still install both solvers, ensuring full
solver-dependent test coverage is maintained.

Also add explicit z3 version confirmation steps to the KNOWNBUG and
THOROUGH jobs, which were previously missing them. This ensures that
jobs expecting z3 will fail loudly if it is accidentally not available.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>
